File: src/primitives/thread.ts
/**
 * Thread Primitive
 *
 * Threads enable focused sub-conversations branching off from a main message.
 * They keep channels clean while allowing deep dives on specific topics.
 */

import { Message } from './message.js';

/**
 * A thread of replies to a message.
 */
export class Thread {
  readonly id: string;
  readonly parentMessageId: string;
  readonly channelId: string;
  readonly createdAt: Date;

  private readonly _replies: Message[] = [];
  private readonly _participantIds: Set<string> = new Set();
  private readonly _followerIds: Set<string> = new Set();

  lastReplyAt?: Date;

  constructor(data: { parentMessageId: string; channelId: string; creatorId: string }) {
    // Thread ID matches parent message ID for easy lookup
    this.id = data.parentMessageId;
    this.parentMessageId = data.parentMessageId;
    this.channelId = data.channelId;
    this.createdAt = new Date();
    // Creator is first participant
    this._participantIds.add(data.creatorId);
  }

  get replyCount(): number {
    return this._replies.length;
  }

  get isActive(): boolean {
    return this.replyCount > 0;
  }

  get participantIds(): string[] {
    return Array.from(this._participantIds);
  }

  get followerIds(): string[] {
    return Array.from(this._followerIds);
  }

  /**
   * Add a reply to the thread.
   */
  addReply(message: Message): Message {
    const reply = new Message({
      ...message.toJSON(),
      threadId: this.id,
      channelId: this.channelId,
    });

    this._replies.push(reply);
    this.lastReplyAt = reply.createdAt;
    this._participantIds.add(reply.senderId);

    return reply;
  }

  /**
   * Get replies in the thread.
   */
  getReplies(options: { limit?: number; before?: Date; after?: Date } = {}): Message[] {
    const { limit = 100, before, after } = options;

    let replies = [...this._replies];

    if (after) {
      replies = replies.filter((r) => r.createdAt > after);
    }
    if (before) {
      replies = replies.filter((r) => r.createdAt < before);
    }

    replies.sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime());
    return replies.slice(0, limit);
  }

  /**
   * Follow a thread for notifications.
   */
  follow(agentId: string): boolean {
    if (!this._followerIds.has(agentId)) {
      this._followerIds.add(agentId);
      return true;
    }
    return false;
  }

  /**
   * Stop following a thread.
   */
  unfollow(agentId: string): boolean {
    return this._followerIds.delete(agentId);
  }

  /**
   * Check if an agent should be notified of thread activity.
   */
  shouldNotify(agentId: string): boolean {
    return this._participantIds.has(agentId) || this._followerIds.has(agentId);
  }

  /**
   * Get thread context for an agent catching up.
   */
  getContext(): {
    threadId: string;
    parentMessageId: string;
    channelId: string;
    replyCount: number;
    participantCount: number;
    participants: string[];
    lastReplyAt: Date | undefined;
  } {
    return {
      threadId: this.id,
      parentMessageId: this.parentMessageId,
      channelId: this.channelId,
      replyCount: this.replyCount,
      participantCount: this._participantIds.size,
      participants: this.participantIds,
      lastReplyAt: this.lastReplyAt,
    };
  }

  toJSON() {
    return {
      id: this.id,
      parentMessageId: this.parentMessageId,
      channelId: this.channelId,
      replyCount: this.replyCount,
      participants: this.participantIds,
      followers: this.followerIds,
      createdAt: this.createdAt,
      lastReplyAt: this.lastReplyAt,
    };
  }

  toString(): string {
    return `Thread(${this.id.slice(0, 8)}): ${this.replyCount} replies, ${this._participantIds.size} participants`;
  }

  *[Symbol.iterator](): Iterator<Message> {
    yield* this._replies;
  }
}

/**
 * Manages threads within a channel.
 */
export class ThreadManager {
  private readonly _threads: Map<string, Thread> = new Map();
  private readonly _messageThreads: Map<string, string> = new Map(); // messageId -> threadId

  /**
   * Get existing thread or create new one for a message.
   */
  getOrCreateThread(parentMessage: Message): Thread {
    const existingThreadId = this._messageThreads.get(parentMessage.id);
    if (existingThreadId) {
      return this._threads.get(existingThreadId)!;
    }

    const thread = new Thread({
      parentMessageId: parentMessage.id,
      channelId: parentMessage.channelId,
      creatorId: parentMessage.senderId,
    });

    this._threads.set(thread.id, thread);
    this._messageThreads.set(parentMessage.id, thread.id);

    return thread;
  }

  /**
   * Get a thread by ID.
   */
  getThread(threadId: string): Thread | undefined {
    return this._threads.get(threadId);
  }

  /**
   * Get the thread associated with a message.
   */
  getThreadForMessage(messageId: string): Thread | undefined {
    const threadId = this._messageThreads.get(messageId);
    return threadId ? this._threads.get(threadId) : undefined;
  }

  /**
   * List all threads, optionally filtered.
   */
  listThreads(options: { channelId?: string; activeOnly?: boolean } = {}): Thread[] {
    const { channelId, activeOnly = false } = options;

    let threads = Array.from(this._threads.values());

    if (channelId) {
      threads = threads.filter((t) => t.channelId === channelId);
    }

    if (activeOnly) {
      threads = threads.filter((t) => t.isActive);
    }

    return threads.sort((a, b) => {
      const aTime = a.lastReplyAt?.getTime() ?? a.createdAt.getTime();
      const bTime = b.lastReplyAt?.getTime() ?? b.createdAt.getTime();
      return bTime - aTime;
    });
  }
}

File: agent_slack/primitives/thread.py
"""
Thread Primitive

Threads enable focused sub-conversations branching off from a main message.
They keep channels clean while allowing deep dives on specific topics.
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Iterator
import uuid

from .message import Message


@dataclass
class Thread:
    """
    A thread of replies to a message.
    
    Key concepts:
    - Threads branch off from a parent message
    - Replies in a thread don't clutter the main channel
    - Threads track participants and activity
    - The parent message "owns" the thread
    
    Design decisions:
    - Thread ID matches parent message ID (1:1 relationship)
    - Replies are ordered chronologically
    - Thread can be "followed" for notifications even without participating
    """
    
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    
    # Parent message that started the thread
    parent_message_id: str = ""
    channel_id: str = ""
    
    # Thread state
    reply_count: int = 0
    participant_ids: set[str] = field(default_factory=set)
    follower_ids: set[str] = field(default_factory=set)  # Watching but not replied
    
    # Timestamps
    created_at: datetime = field(default_factory=datetime.utcnow)
    last_reply_at: Optional[datetime] = None
    
    # Replies (in-memory storage)
    _replies: list[Message] = field(default_factory=list, repr=False)
    
    @property
    def is_active(self) -> bool:
        """Check if thread has any replies."""
        return self.reply_count > 0
    
    @property
    def participants(self) -> list[str]:
        """Get list of participant IDs."""
        return list(self.participant_ids)
    
    def add_reply(self, message: Message) -> Message:
        """
        Add a reply to the thread.
        
        Updates the message to reference this thread and tracks participation.
        """
        message.thread_id = self.id
        message.channel_id = self.channel_id
        
        self._replies.append(message)
        self.reply_count += 1
        self.last_reply_at = message.created_at
        self.participant_ids.add(message.sender_id)
        
        return message
    
    def get_replies(
        self,
        limit: int = 100,
        before: Optional[datetime] = None,
        after: Optional[datetime] = None,
    ) -> list[Message]:
        """
        Get replies in the thread.
        
        Returns replies in chronological order.
        """
        replies = self._replies
        
        if after:
            replies = [r for r in replies if r.created_at > after]
        if before:
            replies = [r for r in replies if r.created_at < before]
        
        replies = sorted(replies, key=lambda m: m.created_at)
        return replies[:limit]
    
    def follow(self, agent_id: str) -> bool:
        """
        Follow a thread for notifications.
        
        Followers get notified of new replies without having to participate.
        """
        if agent_id not in self.follower_ids:
            self.follower_ids.add(agent_id)
            return True
        return False
    
    def unfollow(self, agent_id: str) -> bool:
        """Stop following a thread."""
        if agent_id in self.follower_ids:
            self.follower_ids.remove(agent_id)
            return True
        return False
    
    def should_notify(self, agent_id: str) -> bool:
        """Check if an agent should be notified of thread activity."""
        return agent_id in self.participant_ids or agent_id in self.follower_ids
    
    def get_context(self, include_parent: bool = True) -> dict:
        """
        Get thread context for an agent catching up.
        
        Returns summary information about the thread.
        """
        return {
            "thread_id": self.id,
            "parent_message_id": self.parent_message_id,
            "channel_id": self.channel_id,
            "reply_count": self.reply_count,
            "participant_count": len(self.participant_ids),
            "participants": list(self.participant_ids),
            "last_reply_at": self.last_reply_at.isoformat() if self.last_reply_at else None,
        }
    
    def to_dict(self) -> dict:
        """Serialize thread to dictionary."""
        return {
            "id": self.id,
            "parent_message_id": self.parent_message_id,
            "channel_id": self.channel_id,
            "reply_count": self.reply_count,
            "participants": list(self.participant_ids),
            "followers": list(self.follower_ids),
            "created_at": self.created_at.isoformat(),
            "last_reply_at": self.last_reply_at.isoformat() if self.last_reply_at else None,
        }
    
    def __str__(self) -> str:
        """Human-readable representation."""
        return f"Thread({self.id[:8]}): {self.reply_count} replies, {len(self.participant_ids)} participants"
    
    def __iter__(self) -> Iterator[Message]:
        """Iterate over replies."""
        return iter(self._replies)


class ThreadManager:
    """
    Manages threads within a channel.
    
    This is typically owned by a Channel or Workspace to handle thread
    creation and lookup.
    """
    
    def __init__(self):
        self._threads: dict[str, Thread] = {}  # thread_id -> Thread
        self._message_threads: dict[str, str] = {}  # message_id -> thread_id
    
    def get_or_create_thread(self, parent_message: Message) -> Thread:
        """
        Get existing thread or create new one for a message.
        
        Thread ID matches the parent message ID for easy lookup.
        """
        if parent_message.id in self._message_threads:
            thread_id = self._message_threads[parent_message.id]
            return self._threads[thread_id]
        
        thread = Thread(
            id=parent_message.id,  # Thread ID = Parent message ID
            parent_message_id=parent_message.id,
            channel_id=parent_message.channel_id,
        )
        
        # Parent message author is first participant
        thread.participant_ids.add(parent_message.sender_id)
        
        self._threads[thread.id] = thread
        self._message_threads[parent_message.id] = thread.id
        
        return thread
    
    def get_thread(self, thread_id: str) -> Optional[Thread]:
        """Get a thread by ID."""
        return self._threads.get(thread_id)
    
    def get_thread_for_message(self, message_id: str) -> Optional[Thread]:
        """Get the thread associated with a message."""
        thread_id = self._message_threads.get(message_id)
        if thread_id:
            return self._threads.get(thread_id)
        return None
    
    def list_threads(
        self,
        channel_id: Optional[str] = None,
        active_only: bool = False,
    ) -> list[Thread]:
        """List all threads, optionally filtered."""
        threads = list(self._threads.values())
        
        if channel_id:
            threads = [t for t in threads if t.channel_id == channel_id]
        
        if active_only:
            threads = [t for t in threads if t.is_active]
        
        return sorted(threads, key=lambda t: t.last_reply_at or t.created_at, reverse=True)
